src/VCS/Adapter.php

```diff
@@ -0,0 +1,395 @@
+<?php
+
+namespace Utopia\VCS;
+
+use Exception;
+
+abstract class Adapter
+{
+    public const METHOD_GET = 'GET';
+
+    public const METHOD_POST = 'POST';
+
+    public const METHOD_PUT = 'PUT';
+
+    public const METHOD_PATCH = 'PATCH';
+
+    public const METHOD_DELETE = 'DELETE';
+
+    public const METHOD_HEAD = 'HEAD';
+
+    public const METHOD_OPTIONS = 'OPTIONS';
+
+    public const METHOD_CONNECT = 'CONNECT';
+
+    public const METHOD_TRACE = 'TRACE';
+
+    protected bool $selfSigned = true;
+
+    protected string $endpoint;
+
+    /**
+     * Global Headers
+     *
+     * @var array<string, string>
+     */
+    protected $headers = [];
+
+    /**
+     * Get Adapter Name
+     *
+     * @return string
+     */
+    abstract public function getName(): string;
+
+    /**
+     * Get Adapter Type
+     *
+     * @return string
+     */
+    abstract public function getType(): string;
+
+    /**
+     * Initialize Variables
+     *
+     * @param string $installationId
+     * @param string $privateKey
+     * @param string $githubAppId
+     * @return void
+     */
+    abstract public function initializeVariables(string $installationId, string $privateKey, string $githubAppId): void;
+
+    /**
+     * Generate Access Token
+     *
+     * @param string $privateKey
+     * @param string $githubAppId
+     * @return void
+     */
+    abstract protected function generateAccessToken(string $privateKey, string $githubAppId): void;
+
+    /**
+     * Get user
+     *
+     * @return array<mixed>
+     *
+     */
+    abstract public function getUser(string $username): array;
+
+    /**
+     * Get owner name of the installation
+     *
+     * @return string
+     */
+    abstract public function getOwnerName(string $installationId): string;
+
+    /**
+     * List repositories for Git App
+     * @param int $page page number
+     * @param int $per_page number of results per page
+     * @return array<mixed>
+     */
+    abstract public function listRepositories($page, $per_page): array;
+
+    /**
+     * Get latest opened pull request with specific base branch
+     * @return array<mixed>
+     */
+    abstract public function getBranchPullRequest(string $owner, string $repositoryName, string $branch): array;
+
+    /**
+     * Get repository
+     *
+     * @return array<mixed>
+     */
+    abstract public function getRepository(string $owner, string $repositoryName): array;
+
+    /**
+     * Create new repository
+     *
+     * @return array<mixed> Details of new repository
+     */
+    abstract public function createRepository(string $owner, string $repositoryName, bool $private): array;
+
+    /**
+     * Delete repository
+     */
+    abstract public function deleteRepository(string $owner, string $repositoryName): void;
+
+    /**
+     * Get total repositories count
+     *
+     * @return int
+     */
+    abstract public function getRepositoriesTotalCount(): int;
```

Comment:
```suggestion
    abstract public function getRepositoryCount(): int;
```
Also why doesn't this function take any owner as a parameter?

Reply:
What repository count does it represent

Reply:
lets also group all the repository related functions together in the file, similarly for the pull request related functions etc. Lets have a logical structure in the file

Reply:
We don't use it in Appwrite any longer. So removing it.

Reply:
@christyjacob4 this is fine, it represent the attribute name which is `total_count`